### ISSUE TITLE
`accept-flake-config` regression in 2.14

### ISSUE THREAD

**Issue by amesgen** [NixOS/nix#7955]
**Describe the bug**

Consider a flake that contains a (default) derivation that is cached in a binary cache specified in the `nixConfig` section. Then, assuming it is not already in our store, running `nix build --accept-flake-config` on that flake should not build anything. This has regressed from 2.13 to 2.14.

**Steps To Reproduce**

An example flake for the situation described above is https://github.com/amesgen/stuff/blob/nix-aafc/flake.nix

This will only work on x86_64-linux, as that is what is in the respective cache.

Run
```
nix build --accept-flake-config github:amesgen/stuff/nix-aafc -L
```
(potentially with a `--store /empty-store` flag when running this repeatedly, to make sure that the derivation is not already in your store) and observe that the `hello-variant` derivation is downloaded from `https://cache.amesgen.de/test` for Nix 2.13, but built from source for Nix 2.14 (both 2.14.0 and 2.14.1 as well as current master, which is 0507462c0626d19b59b78e490b7b203df29e7bb0 ATM).

**Expected behavior**

The derivation should be fetched from the cache.

**Bisection**

Bisecting yields that b26562c629604b76c58b789ee520581cc5549433 introduced this bug, pinging @dramforever as the author of #7667.

**Additional context**

This scenario e.g. occurs when using Nix in GitHub Actions CI and setting
```
accept-flake-config = true
```
in `nix.conf` to automatically use the binary caches specified in a flake. With Nix 2.14, this does no longer work; the caches are ignored.

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by nrdxp** [comment]
I'm not sure this is actually fixed or the fix wasn't backported to 2.14. currently stuck on 2.14 due to breaking change on `.drv` files in 2.15 that is gonna require some work in a lot of different places to fix in my stack. I am using 2.14-maintenance, just checked to make sure it is up to date, and still having this issue.

**Comment by amesgen** [comment]
This was not backported, I asked about it in https://github.com/NixOS/nix/pull/7958#issuecomment-1455856547. My patch 77bb950566d4e3b0d98be73a270805ba543f9432 applies cleanly on 2.14-maintenance; it is tiny.

The bug is fixed in 2.15 though.

**Comment by amesgen** [comment]
#7958 was just backported to 2.14-maintenance in #8266.
